src/utils/middleware.ts

```diff
@@ -0,0 +1,205 @@
+import type {
+  JSONRPCMessage,
+  JSONRPCRequest,
+  JSONRPCResponse,
+  JSONRPCResponseResult,
+  MiddlewareFactory,
+} from '../types';
+import { TransformStream } from 'stream/web';
+import { JSONParser } from '@streamparser/json';
+import * as rpcUtils from './utils';
+import { promise } from './utils';
+import * as rpcErrors from '../errors/errors';
+
+/**
+ * This function is a factory to create a TransformStream that will
+ * transform a `Uint8Array` stream to a JSONRPC message stream.
+ * The parsed messages will be validated with the provided messageParser, this
+ * also infers the type of the stream output.
+ * @param messageParser - Validates the JSONRPC messages, so you can select for a
+ *  specific type of message
+ * @param bufferByteLimit - sets the number of bytes buffered before throwing an
+ *  error. This is used to avoid infinitely buffering the input.
+ */
+function binaryToJsonMessageStream<T extends JSONRPCMessage>(
+  messageParser: (message: unknown) => T,
+  bufferByteLimit: number = 1024 * 1024,
+): TransformStream<Uint8Array, T> {
+  const parser = new JSONParser({
+    separator: '',
+    paths: ['$'],
+  });
+  let bytesWritten: number = 0;
+
+  return new TransformStream<Uint8Array, T>({
+    flush: async (controller) => {
+      const waitP = promise();
+      parser.onEnd = () => waitP.resolveP();
+      parser.end();
+      await waitP.p;
+    },
+    start: (controller) => {
+      parser.onValue = (value) => {
+        const jsonMessage = messageParser(value.value);
+        const jsonMessageString = JSON.stringify(jsonMessage);
+        const jsonMessageWithNewline = jsonMessageString + '\n';
+        const jsonMessageObject: T = JSON.parse(jsonMessageWithNewline) as T;
+        controller.enqueue(jsonMessageObject);
+        bytesWritten = 0;
+      };
```

Comment:
I don't understand what's going on here. Why are you adding a `\n` here? Also doing a stringify and parse within the stream here is really expensive.

Reply:
Was this done to help solve #1? If so, this would not be correct way of solving this problem.

Reply:
it was

Reply:
Can you answer @tegefaulkes comment above?

Reply:
> I don't understand what's going on here. Why are you adding a `\n` here? Also doing a stringify and parse within the stream here is really expensive.

Serialises the json to string, adds a new line (wherein it makes it human readable per the bare description of the issue) and then deserialises it back to an object. I now see why it might introduce a lot of overhead, lemme see how else I can resolve it. Maybe before it is sent as a stream would be a good place to change it

Reply:
Ok, I see that the issue is not clear. It does depend on a bunch of context and previous discussion. I'll expand on it in #1 in a moment.